src/core/qgsvirtuallayertask.h

```diff
@@ -0,0 +1,71 @@
+/***************************************************************************
+                          qgsvirtuallayertask.h  -  description
+                             -------------------
+    begin                : Jan 19, 2018
+    copyright            : (C) 2017 by Paul Blottiere
+    email                : blottiere.paul@gmail.com
+ ***************************************************************************/
+
+/***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#ifndef QGSVIRTUALLAYERTASK_H
+#define QGSVIRTUALLAYERTASK_H
+
+#include "qgsvectorlayer.h"
+#include "qgsvirtuallayerdefinition.h"
+#include "qgstaskmanager.h"
+
+/**
+ * \ingroup core
+ *
+ * Initializes a virtual layer with postpone mode activated and reloads the
+ * data in a separated thread.
+ *
+ * \since QGIS 3.2
+ */
+class CORE_EXPORT QgsVirtualLayerTask : public QgsTask
+{
+    Q_OBJECT
+
+  public:
+
+    /**
+     * Constructor.
+     * \param definition The definition to use for initializing the virtual layer
+     */
+    QgsVirtualLayerTask( const QgsVirtualLayerDefinition &definition );
+
+    /**
+     * Returns the underlying virtual layer.
+     */
+    QgsVectorLayer *layer();
```

Comment:
Does this also need a method to "take" the layer? At the moment it retains ownership, so will always be deleted after the task completes. I think having a way to take the result layer would make this class much more useful.

Reply:
Good point! I added a `takeLayer` method.